### PR TITLE
chore: remove `TypeCheckError::Unstructured`

### PR DESCRIPTION
**Pull request by f01dab1e**
# Description

remove `RuntimeErrorKind::{Spanless, UnstructuredError}`

## Problem

Towards #11 

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by f01dab1e** [comment]
@jfecher

**Comment by jfecher** [comment]
@f01dab1e is this just a draft or is it ready for review?

File: crates/noirc_frontend/src/hir/type_check/errors.rs [hunk 1]
```diff
@@ -5,6 +5,25 @@ use thiserror::Error;
...
+    #[error("Binary")]
+    Binary,
+    #[error("Assignment")]
+    Assignment,
+    #[error("Array")]
+    Array,
+    #[error("Array2")]
+    Array2,
+    #[error("Array2")]
+    String,
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/errors.rs:21]:
```suggestion
    #[error(ArrayElements)]
    ArrayElements,
    #[error("ArrayLen")]
    ArrayLen,
    #[error("StringLen")]
    StringLen,
```

File: crates/noirc_frontend/src/hir/type_check/errors.rs [hunk 2]
```diff
@@ -5,6 +5,25 @@ use thiserror::Error;
...
+    #[error("Comparison")]
+    Comparison,
+    #[error("BinOp")]
```
Comment (jfecher) [inline crates/noirc_frontend/src/hir/type_check/errors.rs:24]: Also, do these really need the Error annotations? Can we remove them?

**Review by jfecher — CHANGES_REQUESTED** [review]

File: crates/noirc_frontend/src/hir/type_check/errors.rs (hunk 2 above)
Comment (f01dab1e) [reply crates/noirc_frontend/src/hir/type_check/errors.rs:24]: i added derive for trait Error due to error:
```
error[E0599]: the method `as_dyn_error` exists for reference `&type_check::errors::Source`, but its trait bounds were not satisfied
  --> crates/noirc_frontend/src/hir/type_check/errors.rs:37:72
   |
11 | pub enum Source {
   | ---------------
   | |
   | doesn't satisfy `type_check::errors::Source: AsDynError<'_>`
   | doesn't satisfy `type_check::errors::Source: StdError`
...
37 |     TypeMismatchWithSource { expected: Type, actual: Type, span: Span, source: Source },
   |                                                                        ^^^^^^ method cannot be called on `&type_check::errors::Source` due to unsatisfied trait bounds
   |
```

**Review by jfecher — APPROVED** [review]
LGTM

**Pull request merged**